include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+        op.vmult(*i, u);
+        v += *i;
+        op.range_vector_memory.free(i);
+      };
+
+      op.Tvmult = [&matrix](Domain &v, const Range &u)
+      {
+        matrix.Tvmult(v,u);
+      };
+
+      op.Tvmult_add = [op](Domain &v, const Range &u)
+      {
+        // Capture op by value to have a valid GrowingVectorMemory object
+        // that is bound to the lifetime of this function object
+
+        Domain *i = op.domain_vector_memory.alloc();
+        op.reinit_domain_vector(*i, /*bool fast =*/true);
+        op.Tvmult(*i, u);
+        v += *i;
+        op.domain_vector_memory.free(i);
+      };
+    }
+  };
+
+} /* namespace */
+
+
+/**
+ * \relates LinearOperator
+ *
+ * A function that encapsulates generic @p matrix objects that act on a
+ * compatible Vector type into a LinearOperator. The LinearOperator object
+ * that is created stores a reference to the matrix object. Thus, @p matrix
+ * must remain a valid reference for the whole lifetime of the
+ * LinearOperator object.
+ *
+ * All changes made on @p matrix after the creation of the LinearOperator
+ * object are reflected by the operator object. For example, it is a valid
+ * procedure to first create a LinearOperator and resize, reassemble the
+ * matrix later.
+ *
+ * The Matrix class in question must provide the following minimal
+ * interface:
+ *
+ * @code
+ * class Matrix
+ * {
+ * public:
+ *   // (type specific) information how to create a Range and Domain vector
+ *   // with appropriate size and internal layout
+ *
+ *   // Application of matrix to vector src, writes the result into dst.
+ *   vmult(VECTOR &dst, const VECTOR &src);
+ *
+ *   // Application of the transpose of matrix to vector src, writes the
+ *   // result into dst. (Depending on the usage of the linear operator
+ *   // class this can be a dummy implementation throwing an error.)
+ *   Tvmult(VECTOR &dst, const VECTOR &src);
+ * };
+ * @endcode
+ *
+ * The following (optional) interface is used if available:
+ *
+ * @code
+ * class Matrix
+ * {
+ * public:
+ *   // Application of matrix to vector src, adds the result to dst.
+ *   vmult_add(VECTOR &dst, const VECTOR &src);
+ *
+ *   // Application of the transpose of matrix to vector src, adds the
+ *   // result to dst.
+ *   Tvmult_add(VECTOR &dst, const VECTOR &src);
+ * };
+ * @endcode
+ *
+ * If the Matrix does not provide <code>vmult_add</code> and
+ * <code>Tvmult_add</code>, they are implemented in terms of
+ * <code>vmult</code> and <code>Tvmult</code> (requiring intermediate
+ * storage).
+ *
+ * @author Matthias Maier, 2015
+ */
+template <typename Range = Vector<double>,
+          typename Domain = Range,
+          typename Matrix>
+LinearOperator<Range, Domain> linop(const Matrix &matrix)
+{
+  // implement with the more generic variant below...
+  return linop<Range, Domain, Matrix, Matrix>(matrix, matrix);
+}
+
+
+/**
+ * \relates LinearOperator
+ *
+ * Variant of above function that takes an Exemplar object @p exemplar as
```

Comment:
`template` :-) Maybe `operator_exemplar` or `operator_template`, along with your explanation that it is an " object that stores knowledge how to reinitialize vectors".